core/src/main/java/org/apache/calcite/sql/validate/IdentifierNamespace.java

```diff
@@ -204,11 +206,42 @@ private SqlValidatorNamespace resolveImpl(SqlIdentifier id) {
    * @return    A placeholder {@code TableNamespace}.
    */
   private SqlValidatorNamespace getPlaceholderNamespace(SqlIdentifier id) {
-    SqlValidatorTable table = RelOptTableImpl.create(
+    CalciteSchema parentSchema = getOrCreateParentSchema(id);
+    SqlIdentifier tableId = id.getComponent(id.names.size() - 1);
+    Table table = new UnknownPlaceholderTable(validator.getTypeFactory());
+    SqlValidatorTable valTable = RelOptTableImpl.create(
         (RelOptSchema) validator.getCatalogReader(),
-        new UnknownRecordType(validator.getTypeFactory()),
-        Arrays.asList(id.toString()), null);
-    return new TableNamespace(validator, table);
+        table.getRowType(validator.getTypeFactory()), id.names, null);
+    parentSchema.add(tableId.toString(), table);
+    return new TableNamespace(validator, valTable);
+  }
+
+  /**
+   * Given an identifier, creates the schema chain corresponding to the prefix
+   * list of the identifier, and adds it to the root schema. Any pre-existing
+   * schemas will be retrieved rather than recreated. As an example, if the
+   * identifier is foo.bar.baz, foo will be added as a subschema to the root
+   * schema, bar will be added as a subschema to foo, and then bar will be
+   * returned as the parent schema of baz.
+   * @param id  the identifier for which the parent schema must be created or
+   *            retrieved.
+   * @return    the immediate parent of the object denoted by the identifer -
+   *            in other words, the schema corresponding to the second to last
+   *            component of the identifier. If the identifier consists of only
+   *            a single name, the root schema will be returned.
+   */
+  public CalciteSchema getOrCreateParentSchema(SqlIdentifier id) {
```

Comment:
Is it possible to move this to SqlValidator? I'm not sure if I'll be able to access this from there otherwise.

Reply:
Ok, just moved it.

Reply:
Awesome, thanks!